stablehlo/dialect/TypeInference.cpp

```diff
@@ -2757,7 +2752,45 @@ LogicalResult inferTriangularSolveOp(
     return emitOptionalError(
         location, "Invalid transpose option value for triangular solve");
 
-  inferredReturnShapes.emplace_back(bType.cast<ShapedType>());
+  auto aBounds = encodingToBounds(aType.getEncoding()).vec();
+  auto bBounds = encodingToBounds(bType.getEncoding()).vec();
+  if (aBounds.empty() || bBounds.empty()) {
+    inferredReturnShapes.emplace_back(bType.cast<ShapedType>());
+    return success();
+  }
+  auto resultShape = bType.getShape().vec();
+  auto resultBounds = encodingToBounds(aType.getEncoding()).vec();
+  for (int64_t i = 0; i < aRank - 2; i++) {
+    auto inferredDimAndBoundOrErr = inferMergedDimAndBound(
+        location, i, aBatchDims[i], bBatchDims[i], aBounds[i], bBounds[i]);
+    if (failed(inferredDimAndBoundOrErr)) return failure();
+    resultShape[i] = (*inferredDimAndBoundOrErr).first;
+    resultBounds[i] = (*inferredDimAndBoundOrErr).second;
+  }
+  if (leftSide) {
+    resultShape[bRank - 2] = bType.getShape()[bRank - 2];
+    resultBounds[bRank - 2] = bBounds[bRank - 2];
+    if (isTransposeOrAdjoint) {
```

Comment:
Why does it matter whether `isTransposeOrAdjoint` is true or false? This doesn't come up in [the spec](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#triangular_solve) or in [shape_inference.cc](https://github.com/tensorflow/tensorflow/blob/6a7da3957b5c2c35ad5b7b92e364068f9814235a/tensorflow/compiler/xla/service/shape_inference.cc#L2006).

Reply:
It doesn't show up in shape_inference.cc and the spec since this change is specific to bounds. TriangularSolveOp (C3) requires the last two dimensions to be square, so it does not need any shape inference. However, TRANSPOSE or ADJOINT both require a transpose of the last two dimensions, and if the bounds differ, they need to be swapped. There are two options, keep the current PR's implementation or find the min bound and set both dimensions to it. I went with this approach since we have not decided on inferring the tightest bound, but ideally, we should just be setting the same minimum bounds.